imbi/endpoints/base.py

```diff
@@ -470,3 +478,133 @@ async def patch(self, *args, **kwargs):
+                                                          typing.Any]]]]
+"""Type signature of the function called by PaginatedRequestHandler
+
+This is something that looks like:
+
+  async def retrieve_data(
+      params: dict[str, typing.Any]) -> list[tuple[datetime.datetime, dict]:
+    ...
+
+"""
+
+
+class PaginatedRequestMixin(web.RequestHandler):
+    """Mix this in to repeated retrieve query results ordered by time
+
+    Call the `fetch_items` method to call a specified query method
+    repeatedly until "token.limit" items have been retrieved.  The
+    query method is called with a well-known set of parameters and
+    returns a list of ("when", "data") tuples.  The query parameters
+    are passed as a dictionary that contains at least the following
+    keys:
+
+    * ``earlier``   the earliest (oldest) record to query for
+    * ``later``     the most recent record to query for
+    * ``remaining`` the maximum number of records to return
+
+    The query function should build a query similar that contains
+    ``... some_column BETWEEN %(earlier)s AND %(later)s ...`` as
+    well as ``... LIMIT %(remaining)s``.  The parameters dictionary
+    contains any additional parameters that are passed into the
+    ``fetch_items`` call.
+
+    Call the `pagination_token_from_request` method to retrieve the
+    pagination token from the current request if one exists.  If
+    the token does not exist, then the concrete implementation needs
+    to generate one to pass into `fetch_items`.
+
+    """
+
+    logger: logging.Logger
+
+    async def fetch_items(self, token: PaginationToken, query: QueryMethod,
+                          time_step: datetime.timedelta,
+                          **initial_params) -> list[dict]:
+        """Call this to iterate over a query function and collect the result
+
+        :param token: controls the pagination process
+        :param query: coroutine function to call to retrieve data
+        :param time_step: amount of time to "step" for each iteration.
+            This controls how much data the "query" function is asked
+            to retrieve.
+        :param initial_params: additional parameters to include when
+            calling `query`
+        :returns: list of calling `query` until `token.limit` items
+            are retrieved
+
+        """
+        buckets = []
+        params = initial_params.copy()
+        params.update({'end': token.start, 'remaining': token.limit})
+        while params['remaining'] > 0 and params['end'] >= token.earliest:
+            params.update({
+                'earlier': params['end'] - time_step,
+                'later': params['end'],
+            })
+            buckets.extend(await query(params))
+            buckets.sort(key=operator.itemgetter(0), reverse=True)
+            params.update({
+                'remaining': token.limit - len(buckets),
+                'end': params['end'] - time_step,
+            })
```

Comment:
You can omit `params['end']` because it isn't used by the query, and as a bookkeeping parameter can be inferred from the other values:

```python
params.update({
    'earlier': token.start - time_step,
    'later': token.start,
    'remaining': token.limit,
})
while params['remaining'] > 0 and params['later'] >= token.earliest:
    buckets.extend(await query(params))
    buckets.sort(key=operator.itemgetter(0), reverse=True)
    params.update({
        'earlier': params['earlier'] - time_step,
        'later': params['earlier'],
        'remaining': token.limit - len(buckets),
    })
```

Reply:
There is also an unlikely edge case here: if an entry has a time that's _exactly_ the `earlier`/`later` value, it will be fetched twice, since the `BETWEEN` clause in postgres is inclusive.

Reply:
the `end` entry is what is passed in the token so `earlier` and `later` are derived from it instead.

Reply:
I'm not sure what you mean? I don't see `end` referenced anywhere but here, and it's acting as a bookkeeping value to know when to stop the loop. But you can infer that condition with the existing values. Note that in your implementation `params['earlier']` and `params['end']` always have the same value when you check `params['end'] >= token.earliest`.